src/validation/data/message.rs

```diff
@@ -0,0 +1,182 @@
+// Copyright 2023 Helsing GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::*;
+
+/// Error converting parsed protobuf fileset into custom representation.
+#[derive(Error, Debug, Diagnostic)]
+#[allow(missing_docs)]
+pub enum MessageError {
+    #[error("field number missing")]
+    FieldNumberMissing,
+    #[error("field name missing")]
+    FieldNameMissing,
+    #[error("field type missing")]
+    FieldTypeMissing,
+}
+
+/// Message definition.
+#[derive(Serialize, Deserialize, Default, Clone, Debug, PartialEq, Eq, Diff)]
+#[diff(attr(
+    #[derive(Debug)]
+    #[allow(missing_docs)]
+))]
+pub struct Message {
+    /// Fields defined in this message.
+    #[serde(deserialize_with = "crate::validation::serde::de_int_key")]
+    pub fields: BTreeMap<i32, Field>,
+}
+
+impl Message {
+    /// Try to create new [`Message`] from [`DescriptorProto`].
+    pub fn new(descriptor: &DescriptorProto) -> Result<Self, MessageError> {
+        let mut message = Message::default();
+
+        for field in &descriptor.field {
+            message.fields.insert(
+                field.number.ok_or(MessageError::FieldNumberMissing)?,
+                Field::new(field)?,
+            );
+        }
+
+        Ok(message)
+    }
+}
+
+/// Field defined in this message.
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, Diff)]
+#[diff(attr(
+    #[derive(Debug)]
+    #[allow(missing_docs)]
+))]
+pub struct Field {
+    /// Name of field.
+    pub name: String,
+    /// Type of field.
+    pub type_: FieldType,
+    /// Label of field.
+    pub label: Option<FieldLabel>,
+    /// Default value.
+    pub default: Option<String>,
+}
+
+impl Field {
+    /// Try to create a new [`Field`] from a [`FieldDescriptorProto`].
+    fn new(descriptor: &FieldDescriptorProto) -> Result<Self, MessageError> {
+        Ok(Self {
+            name: descriptor
+                .name
+                .clone()
+                .ok_or(MessageError::FieldNameMissing)?,
+            type_: match descriptor
```

Comment:
`kind` is a better alternative than `type_` imo

Reply:
Nitpick, closing. Happy to get back to this in a future PR.